vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,330 @@
+# Score-Based Generative Modeling
+#
+# Score-Based Generative Modeling through Stochastic Differential Equations.
+# Yang Song, Jascha Sohl-Dickstein, Diederik P. Kingma, 
+# Abhishek Kumar, Stefano Ermon, and Ben Poole.
+# https://arxiv.org/pdf/2011.13456.pdf
+
+using MLDatasets
+using Flux
+using Flux: @functor, chunk, params
+using Flux.Data: DataLoader
+using Parameters: @with_kw
+using BSON
+using CUDA
+using Images
+using Logging: with_logger
+using ProgressMeter: Progress, next!
+using TensorBoardLogger: TBLogger, tb_overwrite
+using Random
+
+"""
+Projection of Gaussian Noise onto a time vector.
+
+# Notes
+This layer will help embed our random times onto the frequency domain. \n
+W is not trainable and is sampled once upon construction - see assertions below.
+
+# References
+paper-  https://arxiv.org/abs/2006.10739 \n
+layers- https://fluxml.ai/Flux.jl/stable/models/basics/#Building-Layers
+"""
+function GaussianFourierProjection(embed_dim, scale)
+    # Instantiate W once
+    W = randn(Float32, embed_dim ÷ 2) .* scale
+    # Return a function that always references the same W
+    function GaussFourierProject(t)
+        t_proj = t' .* W * Float32(2π)
+        [sin.(t_proj); cos.(t_proj)]
+    end
+end
+
+"""
+Helper function that computes the *standard deviation* of 𝒫₀ₜ(𝘹(𝘵)|𝘹(0)).
+
+# Notes
+Derived from the Stochastic Differential Equation (SDE):    \n
+                𝘥𝘹 = σᵗ𝘥𝘸,      𝘵 ∈ [0, 1]                   \n
+
+We use properties of SDEs to analytically solve for the stddev
+at time t conditioned on the data distribution. \n
+
+We will be using this all over the codebase for computing our model's loss,
+scaling our network output, and even sampling new images!
+"""
+function marginal_prob_std(t, sigma=25.0f0)
+    sqrt.((sigma .^ (2t) .- 1.0f0) ./ 2.0f0 ./ log(sigma))
+end
+
+"""
+Create a UNet architecture as a backbone to a diffusion model. \n
+
+# Notes
+Images stored in WHCN (width, height, channels, batch) order. \n
+In our case, MNIST comes in as (28, 28, 1, batch). \n
+
+# References
+paper-  https://arxiv.org/abs/1505.04597 \n
+Conv-   https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.Conv \n
+TConv-  https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.ConvTranspose \n
+GNorm-  https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.GroupNorm \n
+Flux-   https://fluxml.ai/Flux.jl/stable/models/advanced/#Custom-Model-Example
```

Comment:
I don't think references to Flux layers need to be in the docstrings. External source material is sufficient.
```suggestion
```